decoder/api_pokemon_scan_v2.go

```diff
@@ -0,0 +1,428 @@
+	AtkIv   *ApiPokemonDnfMinMax8 `json:"atk_iv"`
+	DefIv   *ApiPokemonDnfMinMax8 `json:"def_iv"`
+	StaIv   *ApiPokemonDnfMinMax8 `json:"sta_iv"`
+	Level   *ApiPokemonDnfMinMax8 `json:"level"`
+	Cp      *ApiPokemonDnfMinMax  `json:"cp"`
+	Gender  *ApiPokemonDnfMinMax8 `json:"gender"`
+	Size    *ApiPokemonDnfMinMax8 `json:"size"`
+	Little  *ApiPokemonDnfMinMax  `json:"pvp_little"`
+	Great   *ApiPokemonDnfMinMax  `json:"pvp_great"`
+	Ultra   *ApiPokemonDnfMinMax  `json:"pvp_ultra"`
+}
+
+type ApiPokemonDnfId struct {
+	Pokemon int16  `json:"id"`
+	Form    *int16 `json:"form"`
+}
+
+type ApiPokemonDnfMinMax struct {
+	Min int16 `json:"min"`
+	Max int16 `json:"max"`
+}
+
+type ApiPokemonDnfMinMax8 struct {
+	Min int8 `json:"min"`
+	Max int8 `json:"max"`
+}
+
+func internalGetPokemonInArea2(retrieveParameters ApiPokemonScan2) []uint64 {
+	type dnfFilterLookup struct {
+		pokemon int16
+		form    int16
+	}
+
+	dnfFilters := make(map[dnfFilterLookup][]ApiPokemonDnfFilter)
+
+	for _, filter := range retrieveParameters.DnfFilters {
+		if len(filter.Pokemon) > 0 {
+			for _, keyString := range filter.Pokemon {
+				pokemonId := keyString.Pokemon
+				if pokemonId == 0 {
+					pokemonId = -1
+				}
+				var formId int16 = -1
+				if keyString.Form != nil {
+					formId = *keyString.Form
+				}
+				key := dnfFilterLookup{
+					pokemon: pokemonId,
+					form:    formId,
+				}
+				dnfFilters[key] = append(dnfFilters[key], filter)
+			}
+		} else {
+			key := dnfFilterLookup{
+				pokemon: -1,
+				form:    -1,
+			}
+			dnfFilters[key] = append(dnfFilters[key], filter)
+		}
+	}
+
+	start := time.Now()
+
+	minLocation := retrieveParameters.Min
+	maxLocation := retrieveParameters.Max
+
+	maxPokemon := config.Config.Tuning.MaxPokemonResults
+	if retrieveParameters.Limit > 0 && retrieveParameters.Limit < maxPokemon {
+		maxPokemon = retrieveParameters.Limit
+	}
+
+	pokemonExamined := 0
+	pokemonSkipped := 0
+
+	isPokemonDnfMatch := func(pokemonLookup *PokemonLookup, pvpLookup *PokemonPvpLookup, filter *ApiPokemonDnfFilter) bool {
+		if filter.Iv != nil && (pokemonLookup.Iv < filter.Iv.Min || pokemonLookup.Iv > filter.Iv.Max) ||
+			filter.StaIv != nil && (pokemonLookup.Sta < filter.StaIv.Min || pokemonLookup.Sta > filter.StaIv.Max) ||
+			filter.AtkIv != nil && (pokemonLookup.Atk < filter.AtkIv.Min || pokemonLookup.Atk > filter.AtkIv.Max) ||
+			filter.DefIv != nil && (pokemonLookup.Def < filter.DefIv.Min || pokemonLookup.Def > filter.DefIv.Max) ||
+			filter.Level != nil && (pokemonLookup.Level < filter.Level.Min || pokemonLookup.Level > filter.Level.Max) ||
+			filter.Cp != nil && (pokemonLookup.Cp < filter.Cp.Min || pokemonLookup.Cp > filter.Cp.Max) ||
+			filter.Gender != nil && (pokemonLookup.Gender < filter.Gender.Min || pokemonLookup.Gender > filter.Gender.Max) ||
+			filter.Size != nil && (pokemonLookup.Size < filter.Size.Min || pokemonLookup.Size > filter.Size.Max) {
+			return false
+		}
+
+		if pvpLookup == nil ||
```

Comment:
Should be `pvpLookup != nil && (...)`

Reply:
https://github.com/UnownHash/Golbat/pull/134/files#diff-599121fdc02df91e88c5bf7eceb2a14ce6e65b7312ce0d7c8559a0cb6921e3a6R627

Reply:
Actually it should be
```go
        if filter.Little != nil && (pvpLookup == nil || pvpLookup.Little < filter.Little.Min || pvpLookup.Little > filter.Little.Max) ||
            filter.Great != nil && (pvpLookup == nil || pvpLookup.Great < filter.Great.Min || pvpLookup.Great > filter.Great.Max) ||
            filter.Ultra != nil && (pvpLookup == nil || pvpLookup.Ultra < filter.Ultra.Min || pvpLookup.Ultra > filter.Ultra.Max) {
            return false
        }
```

What's the point of having a separate `pvpLookup` object anyways? Completely unnecessary complexity. 🤷

Reply:
Good catch, have pushed.
The pvpLookup object is rebuilt only when ohbem data is available (which is not every update). The core lookup data is rebuilt every time anything else changes.  
Originally it was more complex and generic - but through testing this was found to substantially impact performance with only a theoretical future benefit (if they introduce new leagues, for example).